### PR TITLE
REF: Refactor Datetimelike delegation

### PR DESCRIPTION
**Pull request by TomAugspurger**
This is a generalization of PeriodIndex's dispatching to
PeriodArray, without any actual changes yet. This is split
from #24024, where DatetimeIndex and TimedeltaIndex will
implement and inherit from delgates similiar to PeriodDelegateMixin.

### CODE REVIEW

**Comment by TomAugspurger** [comment]
These latest two commits have some actual changes:

1. Removes "delegation" for size, shape, and asi8. These were being overwritten anyway (I think).
2. Adds delegation for `asfreq` and `to_timestamp`. I don't understand my earlier comment about why they couldn't be dispatched normally (but perhaps tests will turn something up).

File: pandas/core/indexes/datetimelike.py [hunk 1]
```diff
@@ -855,3 +857,48 @@ def f(self):
     f.__name__ = fget.__name__
     f.__doc__ = fget.__doc__
     return property(f)
+
+
+class DatetimelikeDelegateMixin(PandasDelegate):
```
Comment (jbrockmendel) [inline pandas/core/indexes/datetimelike.py]: Is PandasDelegate overkill for this situation?  It's not like `Series.dt` where we construct an object and need to cache it.  Couldn't we just use something like:

```
def pass_through_methods(names):
    def decorator(cls):
        for name in names:
            method = wrap_array_method(...)
            setattr(cls, name, method)
        return cls
```

There's definitely overlap with the PandasDelegate machinery, but I think that machinery makes for really dense reading, so should be reserved for when we really need it.

File: pandas/core/indexes/datetimelike.py [hunk 2]
```diff
@@ -855,3 +857,48 @@ def f(self):
...
+            result = Index(result, name=self.name)
+        return result
+
+    def _delegate_property_set(self, name, value, *args, **kwargs):
+        setattr(self._data, name, value)
+
+    def _delegate_method(self, name, *args, **kwargs):
+        result = operator.methodcaller(name, *args, **kwargs)(self._data)
+        if name not in self._raw_methods:
+            result = Index(result, name=self.name)
```
[the remark this answers is not in the document]
Comment (TomAugspurger) [reply pandas/core/indexes/datetimelike.py]: I think all the delegated methods return a single value.

File: pandas/core/indexes/datetimelike.py (hunk 1 above)
Comment (TomAugspurger) [reply pandas/core/indexes/datetimelike.py]: Perhaps you're thinking of `Properties` (which subclasses `PandasDelegate`)? `PandasDelegate` doesn't have an `__init__`, it just handles registering and creating the delegated properties.

Comment (TomAugspurger) [reply pandas/core/indexes/datetimelike.py]: But, I don't have a strong preference of one of the other. We use `PandasDelegate` on master, for period, so this PR is a strict generalization of that.

File: pandas/core/indexes/datetimelike.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (jbrockmendel) [reply pandas/core/indexes/datetimelike.py]: If #23885 goes through that will change

[the remark this answers is not in the document]
Comment (TomAugspurger) [reply pandas/core/indexes/datetimelike.py]: Gotcha. I'd prefer to wait until we can hit that code path before worrying about it.

**Comment by jreback** [comment]
lgtm. merge when ready @TomAugspurger @jbrockmendel

File: pandas/core/indexes/datetimelike.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (jbrockmendel) [reply pandas/core/indexes/datetimelike.py]: Totally fair.

**Comment by jbrockmendel** [comment]
LGTM.  I might poke at simplifying down the road, but that shouldn't be a blocker.

**Comment by jreback** [comment]
ok I merged the other one, so @TomAugspurger if you can rebase

**Comment by jreback** [comment]
thanks @TomAugspurger

**Pull request merged**